cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -1788,134 +1396,97 @@ TEST_F(TestCast, IdentityCasts) {
-                         UInt32Type, UInt64Type, Int64Type, FloatType, DoubleType,
-                         Date32Type, Date64Type, FixedSizeBinaryType, BinaryType>
-    TestTypes;
-
-TYPED_TEST_SUITE(TestNullCast, TestTypes);
-
-TYPED_TEST(TestNullCast, FromNull) {
-  // Null casts to everything
-  const int length = 10;
-
-  // Hack to get a DataType including for parametric types
-  std::shared_ptr<DataType> out_type =
-      TestBase::MakeRandomArray<typename TypeTraits<TypeParam>::ArrayType>(0, 0)->type();
-
-  NullArray arr(length);
-
-  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> result, Cast(arr, out_type));
-  ASSERT_OK(result->ValidateFull());
-
-  ASSERT_TRUE(result->type()->Equals(*out_type));
-  ASSERT_EQ(length, result->length());
-  ASSERT_EQ(length, result->null_count());
+TEST(Cast, FromNull) {
+  for (auto to_type : {
+           null(),
+           uint8(),
+           int8(),
+           uint16(),
+           int16(),
+           uint32(),
+           int32(),
+           uint64(),
+           int64(),
+           float32(),
+           float64(),
+           date32(),
+           date64(),
+           fixed_size_binary(10),
+           binary(),
+           utf8(),
+       }) {
+    ASSERT_OK_AND_ASSIGN(auto expected, MakeArrayOfNull(to_type, 10));
+    CheckCast(std::make_shared<NullArray>(10), expected);
+  }
 }
 
 // ----------------------------------------------------------------------
 // Test casting from DictionaryType
 
-template <typename TestType>
-class TestDictionaryCast : public TestCast {};
-
-typedef ::testing::Types<NullType, UInt8Type, Int8Type, UInt16Type, Int16Type, Int32Type,
-                         UInt32Type, UInt64Type, Int64Type, FloatType, DoubleType,
-                         Date32Type, Date64Type, FixedSizeBinaryType, BinaryType>
-    TestTypes;
+TEST(Cast, FromDictionary) {
+  ArrayVector dictionaries;
+  dictionaries.push_back(std::make_shared<NullArray>(5));
 
-TYPED_TEST_SUITE(TestDictionaryCast, TestTypes);
+  for (auto num_type : kNumericTypes) {
+    dictionaries.push_back(ArrayFromJSON(num_type, "[23, 12, 45, 12, null]"));
+  }
 
-TYPED_TEST(TestDictionaryCast, Basic) {
-  std::shared_ptr<Array> dict =
-      TestBase::MakeRandomArray<typename TypeTraits<TypeParam>::ArrayType>(5, 1);
-  for (auto index_ty : all_dictionary_index_types()) {
-    auto indices = ArrayFromJSON(index_ty, "[4, 0, 1, 2, 0, 4, null, 2]");
-    auto dict_ty = dictionary(index_ty, dict->type());
-    auto dict_arr = *DictionaryArray::FromArrays(dict_ty, indices, dict);
-    std::shared_ptr<Array> expected = *Take(*dict, *indices);
+  for (auto string_type : kBaseBinaryTypes) {
+    dictionaries.push_back(
+        ArrayFromJSON(string_type, R"(["foo", "bar", "baz", "foo", null])"));
+  }
 
-    this->CheckPass(*dict_arr, *expected, expected->type(), CastOptions::Safe(),
-                    /*check_scalar=*/false);
+  for (auto dict : dictionaries) {
+    for (auto index_type : all_dictionary_index_types()) {
```

Comment:
Nit: should we have `kDictionaryIndexTypes`?

Reply:
sure